test/object-store/sync/flx_sync.cpp

```diff
@@ -287,6 +287,70 @@ TEST_CASE("app: error handling integration test", "[sync][flx][baas]") {
     config.sync_config->error_handler = std::move(error_handler);
     config.sync_config->client_resync_mode = ClientResyncMode::Manual;
 
+    SECTION("Resuming while waiting for session to auto-resume") {
+        enum class BarrierState { InitialSuspend, InitialResume, SecondBind, SecondSuspend, SecondResume, Done };
+        TestingStateMachine<BarrierState> state(BarrierState::InitialSuspend);
+        config.sync_config->on_sync_client_event_hook = [&](std::weak_ptr<SyncSession>,
+                                                            const SyncClientHookData& data) {
+            if (data.event == SyncClientHookEvent::BindMessageSent) {
+                state.transition(BarrierState::SecondBind, BarrierState::SecondSuspend);
+            }
+            if (data.event != SyncClientHookEvent::SessionSuspended) {
+                return SyncClientHookAction::NoAction;
+            }
+
+            if (state.transition(BarrierState::InitialSuspend, BarrierState::InitialResume) ==
+                BarrierState::InitialSuspend) {
+                state.wait_for(BarrierState::SecondBind);
+            }
+            else if (state.transition(BarrierState::SecondSuspend, BarrierState::SecondResume) ==
+                     BarrierState::SecondSuspend) {
+                state.wait_for(BarrierState::Done);
+            }
+            return SyncClientHookAction::NoAction;
+        };
+        auto r = Realm::get_shared_realm(config);
+        wait_for_upload(*r);
+        nlohmann::json error_body = {
+            {"tryAgain", true},           {"message", "fake error"},
+            {"shouldClientReset", false}, {"isRecoveryModeDisabled", false},
+            {"action", "Transient"},      {"backoffIntervalSec", 900},
+            {"backoffMaxDelaySec", 900},  {"backoffMultiplier", 1},
+        };
+        nlohmann::json test_command = {{"command", "ECHO_ERROR"},
+                                       {"args", nlohmann::json{{"errorCode", 229}, {"errorBody", error_body}}}};
+
+        // First we trigger a retryable transient error that should cause the client to try to resume the
+        // session in 5 minutes.
+        auto test_cmd_res =
+            wait_for_future(SyncSession::OnlyForTesting::send_test_command(*r->sync_session(), test_command.dump()))
+                .get();
+        REQUIRE(test_cmd_res == "{}");
+
+        state.wait_for(BarrierState::InitialResume);
+
+        // Once we're suspended, immediately tell the sync client to resume the session. This should cancel the
+        // timer that would have auto-resumed the session.
+        r->sync_session()->handle_reconnect();
+        state.transition(BarrierState::InitialResume, BarrierState::SecondBind);
```

Comment:
This is noop if 'from' is not the first arg. It's hard to reason about just looking at the code. Why not throw instead?

Reply:
What do you mean by "throw"? Throw where? I could assert that we're in the initial resume state here. I wish we had a better generic way of building little state machines where these kinds of tests have to transition between a bunch of states to be deterministic.

Reply:
If you make transition to return a bool (as suggested), you could just check the return value is true.

Reply:
I agree it's a bit hard to follow the code with all the states. I played a bit with it and I think it's enough to have a bool and toggle its value (i.e, the test thread waits for true while the event loop thread waits for false).

Reply:
Yes, that's what i meant: throw and return bool and check it in state.transition. Sorry for not being clear on this. Since you test here quite tricky states transition, it would be great to actually verify in testing code that transition actually happens as expected.

Reply:
I've updated this a bit so we REQUIRE that we're in the right state out in the testing code.